### PR TITLE
Add --format flag to output SPDX document as JSON

### LINKED ISSUE #103: Support additional output formats for SPDX
From https://spdx.dev/resources/use/#documents:

> SPDX Documents can use one of five file formats, tag/value (.spdx), JSON(.spdx.json), YAML(.spdx.yml), RDF/xml(spdx.rdf) and spreadsheets (.xls).

On `bom generate`, would like to be able to specify `--format=json` to get an SPDX document in JSON format.

### PR DESCRIPTION
**Pull request by jdolitsky**
// Description updated by @puerco 

This commit builds on @jdolitsky 's work to enable JSON output in `bom` by combining his SPDX-JSON code with one of my branches implementing a new query and filtering feature. 

## Query Language

The query feature enables adds a query subsystem to bom to enable a new subcommand that allows searching in an sbom. Using the test sbom included in this pr we can now do things like, for example, searching for packages of bash in the sbom:

```
bom document query pkg/spdx/testdata/nginx.spdx 'name:bash'

 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
 bash-5.1-2+b3
```
The initial implementation is rough still but already includes 4 filters (all, purl, depth and name).  Documentation of the query language will come soon

## JSON Output

Building on Josh's work, we now have a native JSON serializer. The serializer uses the query language filters to flatten the packages and files and render them in the JSON package, the output validates now but I (@puerco) need to add a couple of tests to ensure different SBOM sources render correctly.

Generation now takes the flag added by @jdolitsky and writes JSON and Tag-Value. Example:

```
❯ bom generate --format=json . > spdx.json
❯ java -jar tmp/tools-java-1.0.4-jar-with-dependencies.jar Verify spdx.json 
WARNING: sun.reflect.Reflection.getCallerClass is not supported. This will impact performance.
This SPDX Document is valid.

```

Check the attached sample SBOM below for a sample

Resolves #103 

```release-note
- **bom now can write SBOMs in JSON!**
- New `bom document query` subcommand allows querying SBOMs using a new filtering language
```

### CODE REVIEW

File: pkg/spdx/document.go
Comment (jdolitsky) [inline pkg/spdx/document.go:252]: Any of the logic added here will not be included in the JSON ... it seems like these types of relationships should happen ahead of time instead of being printed out as strings, but I don't have a full understanding of SPDX and what that entails

**Review by cpanato — COMMENTED** [review]
/ok-to-test

**Comment by jdolitsky** [comment]
Any issue with importing `github.com/spdx/tools-golang` ?

We could follow this example: https://github.com/spdx/tools-golang/blob/main/examples/9-tvtojson/exampletvtojson.go

**Comment by jdolitsky** [comment]
OK, this works properly now, but it depends on this PR: https://github.com/spdx/tools-golang/pull/139

So CI is failing due to the replace statement in go.mod:
```
go.mod:86:1: replacement are not allowed: github.com/spdx/tools-golang (gomoddirectives)
replace github.com/spdx/tools-golang => github.com/ion-channel/tools-golang v0.0.0-20220425222917-af3d04c69209
^
```

🤷 Not sure where to go from here

**Comment by ianling** [comment]
It may be at least a week or two before we potentially get to merging the PR you linked in the spdx/tools-golang repo. Feel free to import ion-channel/tools-golang temporarily instead of using the `replace` statement if you feel safe doing so (I am the author of the changes in that fork). We use it in production ourselves while we are working on getting my changes merged upstream.

File: pkg/spdx/format.go
Comment (jdolitsky) [inline pkg/spdx/format.go:31]: Tried following @ianling 's advice to import the fork, but that doesn't work without replace statement due to:
```
$ go mod tidy
go: github.com/ion-channel/tools-golang@v0.0.0-20220425222917-af3d04c69209: parsing go.mod:
	module declares its path as: github.com/spdx/tools-golang
	        but was required as: github.com/ion-channel/tools-golang
```

So I forked the fork into a new org called "this-is-a-fork-remove-me-asap", updated the module name, cut a release, and added this TODO

**Comment by puerco** [comment]
@ianling can we help with any of the remaining PRs to help import the module? I'm not comfortable letting in the change pointing to the fork as it goes against the [Kubernetes go dependency guidance](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/godep.md#reviewing-and-approving-dependency-changes) and this is a critical library in the k8s release process.

/cc @lumjjb 
/hold

**Comment by ianling** [comment]
@puerco unfortunately no, they are held up due to the maintainers' lack of available time to review the PRs.

**Comment by puerco** [comment]
@jdolitsky I've rebased your PR on top of my experimental query branch and reworked the SPDX-JSON output using the filtering capabilities of the query language. Basically, the query filter allows selecting elements in the SBOM via certain criteria. So the new JSON serializer works by selecting and flattening all elements (packages and files) and populating the SPDX v2.2.2 type with the data.

We now have this mammoth PR but in the end, bom now has gained two new great features. Thank you so much!

I've edited the PR description to account for the changes.

**Comment by puerco** [comment]
Attaching a sample, valid SBOM in SPDX-JSON :rocket:

[spdx.json.gz](https://github.com/kubernetes-sigs/bom/files/8883367/spdx.json.gz)

**Comment by puerco** [comment]
Thanks @jdolitsky !
/lgmt

**Comment by puerco** [comment]
Thanks @jdolitsky !
/lgtm

**Review by puerco — APPROVED** [review]

**Pull request merged**
